### PR TITLE
Add support for random Generators in Numba backend

### LINKED ISSUE #316: Use Numba Generators for random graphs and deprecate shared RandomState variables
### Description

RandomState are legacy in numpy and we can save some complexity by letting go of them in PyTensor.

We were "obliged" to keep them because that's the only kind of RNG that Numba supported until now

### LINKED ISSUE #568: `RandomVariable` treats `size=()` as `size=None`
### Description

This is different than numpy

```python
import numpy as np
import pytensor.tensor as pt

pt.random.normal([0], size=None).eval()  # array([-0.57354733])
pt.random.normal([0], size=()).eval()  # array([-2.39088062])

np.random.normal([0], size=None)  # array([0.10438181])
np.random.normal([0], size=()).eval()  # ValueError: Output size () is not compatible with broadcast dimensions of inputs (1,).
```

And also JAX (and probably Numba?)

```python
import jax

jax.random.lognormal(jax.random.PRNGKey(0), sigma=jax.numpy.ones((1,)), shape=())
# parameter shapes must be broadcast-compatible with shape argument, 
# and the result of broadcasting the shapes must equal the shape argument, 
# but got result (1,) for shape argument ().
```

### LINKED ISSUE #701: Get rid of dummy type input to RVs
### Description

RandomVariables third input is a numerical integer than indexes into a list of predefined types to indicate if a variable is float64 or float32, ...

This input is useless because PyTensor variables are not allowed to change types at runtime. We should make it a property of the Op. This requires defining Ops dynamically whenever they are initialized, which means there is no canonical `pt.random.normal`. This is not too bad imo and simplifies handling (and visualizing) RV graphs

### PR DESCRIPTION
**Pull request by ricardoV94**
### TODO:
- [x] Merge code duplication into elemwise_codegen
- [x] Handle broadcast via size argument
- [x] Handle more than normal
- [x] Add expand_dims in make_node. This is actually something that ends up being useful in other places, as in vectorization and dim analysis. It also canonicalizes graphs that are equivalent, as the expand_dims is always implict
- [x] Handle non-scalar inputs / outputs. This would also solve Blockwise for Numba backend!
- [x] Find a way to copy RNGs if inplace=False. Not a huge priority, but a nice to have for consistency
- [x] Remove RandomState from codebase. I vote for no deprecation as this is functionality that was broken in Numba anyway (no way to reseed), and non-default for other backends.

## Related Issue

- [x] Closes #316 
- [x] Closes #701
- [x] Closes #568  

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ricardoV94** [comment]
Size and rng copy are now supported

**Comment by ricardoV94** [comment]
@aseyboldt Elemwise is working again in case you want to benchmark before/after

**Comment by aseyboldt** [comment]
Looks like this currently breaks pymc, I'm getting this import error:

```
File [~/git/pymc-dev/pymc/pytensorf.py:46](http://localhost:7890/lab/tree/nuts-py/notebooks/pymc-dev/pymc/pytensorf.py#line=45)
     44 from pytensor.tensor.random.op import RandomVariable
     45 from pytensor.tensor.random.type import RandomType
---> 46 from pytensor.tensor.random.var import (
     47     RandomGeneratorSharedVariable,
     48     RandomStateSharedVariable,
     49 )
     50 from pytensor.tensor.rewriting.shape import ShapeFeature
     51 from pytensor.tensor.sharedvar import SharedVariable, TensorSharedVariable

ImportError: cannot import name 'RandomStateSharedVariable' from 'pytensor.tensor.random.var' ([/home/adr/git/pytensor/pytensor/tensor/random/var.py](http://localhost:7890/lab/tree/nuts-py/notebooks/pytensor/pytensor/tensor/random/var.py))
```

**Comment by aseyboldt** [comment]
I did some benchmarks with the changed elemwise. And I guess the good news is that it is about 400x faster, bad news is that now it doesn't seem to compute the right thing :-)

```python
import pytensor.tensor as pt
import pytensor
import numpy as np

N = 1000
x = pt.vector(shape=(N,))
y = pt.vector(shape=(N,))
z = x[None, :] + 2*y[:, None]

rng = np.random.default_rng(42)
u = rng.normal(size=N)
v = rng.normal(size=N)
pt_func = pytensor.function([x, y], z, mode="NUMBA")
func = pt_func.vm.jit_fn

func(u, v)[0].shape
# (1, 1000)
```
Even though the graph wants to produce a `(1000, 1000)`:

```
Add [id A] <Matrix(float64, shape=(1000, 1000))> 3
 ├─ ExpandDims{axis=0} [id B] <Matrix(float64, shape=(1, 1000))> 2
 │  └─ <Vector(float64, shape=(1000,))> [id C] <Vector(float64, shape=(1000,))>
 └─ Mul [id D] <Matrix(float64, shape=(1000, 1))> 1
    ├─ [[2.]] [id E] <Matrix(float64, shape=(1, 1))>
    └─ ExpandDims{axis=1} [id F] <Matrix(float64, shape=(1000, 1))> 0
       └─ <Vector(float64, shape=(1000,))> [id G] <Vector(float64, shape=(1000,))>
```

Also strange: There doesn't seem to be fusion?

**Comment by ricardoV94** [comment]
> Looks like this currently breaks pymc, I'm getting this import error:
> 
> ```
> File [~/git/pymc-dev/pymc/pytensorf.py:46](http://localhost:7890/lab/tree/nuts-py/notebooks/pymc-dev/pymc/pytensorf.py#line=45)
>      44 from pytensor.tensor.random.op import RandomVariable
>      45 from pytensor.tensor.random.type import RandomType
> ---> 46 from pytensor.tensor.random.var import (
>      47     RandomGeneratorSharedVariable,
>      48     RandomStateSharedVariable,
>      49 )
>      50 from pytensor.tensor.rewriting.shape import ShapeFeature
>      51 from pytensor.tensor.sharedvar import SharedVariable, TensorSharedVariable
> 
> ImportError: cannot import name 'RandomStateSharedVariable' from 'pytensor.tensor.random.var' ([/home/adr/git/pytensor/pytensor/tensor/random/var.py](http://localhost:7890/lab/tree/nuts-py/notebooks/pytensor/pytensor/tensor/random/var.py))
> ```

The label `major` indicates breaking changes, and in PyMC we always pin a specific range, so it won't pick this up when a new version is released. Several other things are breaking changes in this PR including removing the `dtype` fake input and using None instead of () for no size

**Comment by ricardoV94** [comment]
> Also strange: There doesn't seem to be fusion?

There is no fusion when inputs broadcast together.

**Comment by ricardoV94** [comment]
Failing code is funny. No test we had picked it up xD?

**Comment by aseyboldt** [comment]
> There is no fusion when inputs broadcast together.

True, I didn't think that through...

> Edit: I like the speedups, maybe it's an okay trade-off?

Yeah, you gotta make some sacrifices if you want good performance. And that way we can easily do all our computations on the CPU, and don't have buy those expensive nvidia thingies.

**Comment by ricardoV94** [comment]
Funny, enough RVs are broadcasting correctly:

```python
import pytensor.tensor as pt
import pytensor

N = 1000
x = pt.random.normal(size=(1, N,))
y = pt.random.normal(size=(N, 1))
z = pt.random.normal(x, y)

# rng = np.random.default_rng(42)
# u = rng.normal(size=N)
# v = rng.normal(size=N)
pt_func = pytensor.function([], z, mode="NUMBA")

# pytensor.dprint(pt_func, print_type=True)
assert pt_func().shape == (1000, 1000)
```

**Comment by ricardoV94** [comment]
@aseyboldt I think your Elemwise example is failing in main as well?

**Comment by ricardoV94** [comment]
Commit `Refactor encoding helper` introduces the Elemwise bug

**Comment by ricardoV94** [comment]
Ah just something dumb: `output_bc_patterns = tuple([out.type.broadcastable for out in node.inputs])`

**Comment by aseyboldt** [comment]
> Ah just something dumb: output_bc_patterns = tuple([out.type.broadcastable for out in node.inputs])

Let me guess, AI autocomplete? :-)

**Comment by ricardoV94** [comment]
> > Ah just something dumb: output_bc_patterns = tuple([out.type.broadcastable for out in node.inputs])
> 
> Let me guess, AI autocomplete? :-)

Actually no, I changed that on purpose, before it was setting the output bc to False for the outputs, which didn't make sense either? I pushed the fixes

**Comment by ricardoV94** [comment]
I'll need to rebase to get your jit options in as well. Question, the overload needs to include them, because it's part of the dispatch signature?

**Comment by aseyboldt** [comment]
Yes, `overload` calls njit on the returned function, and uses `jit_options` for that njit call.

**Comment by ricardoV94** [comment]
I rebased, the no cython wrapper made the random test file quite faster!

**Comment by ricardoV94** [comment]
> I rebased, the no cython wrapper made the random test file quite faster!

Or maybe not. Anyway doesn't feel too bad

**Comment by ricardoV94** [comment]
This is finally ready for review!

**Comment by ricardoV94** [comment]
For the numba iteration I went with a wrap OFG that adds the `core_shape` as an explicit input. I felt this was the cleanest solution

File: pytensor/link/numba/dispatch/basic.py [hunk 1]
```diff
@@ -65,7 +66,8 @@ def numba_njit(*args, **kwargs):
     # Supress caching warnings
     warnings.filterwarnings(
         "ignore",
-        message='Cannot cache compiled function "numba_funcified_fgraph" as it uses dynamic globals',
+        # FIXME: Specific messages are not working
+        # message='Cannot cache compiled function "store_core_outputs" as it uses dynamic globals (such as ctypes pointers and large global arrays)',
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:70]: I cannot get the warning suppressed with any message whatsoever. The relevant test is 

https://github.com/pymc-devs/pytensor/blob/1e96b8942bc42fd48af50e138c5b2458083c6091/tests/link/numba/test_basic.py#L1071-L1078

Comment (twiecki) [reply pytensor/link/numba/dispatch/basic.py:70]: Wasn't there a pymc issue that we were setting warning levels wrong in a way they couldn't be ignored? I can't find the issue though.

Comment (ricardoV94) [reply pytensor/link/numba/dispatch/basic.py:70]: I think that's different. That was about the logging module, not warnings.

**Review by aseyboldt — APPROVED** [review]
Cool stuff :-)

Higher level the changes are something like this, right? (am I forgetting something important?)

- Add numba implementations for the random variables. (By using existing numba impls or adapting the numpy implementation and writing numba code)
- Change the elemewise to allow passing of extra arguments and to pass output arrays to the core function. This means the code can be used for the blockwise-like random variables. I ran some benchmarks and it seems this didn't make the performance worse.
- Some minor changes to the numba rewrites: Adding rewrites for the inner graph of scans. (I thought we had a flag for ops that show if they have an inner graph and trigger them through that somehow? But I couldn't find that code)
- Change random variables so that they specify a signature. Maybe this should be mentioned in the release notes because it affects for instance pymc implementations or custom distributions?
- Introduce a numba rewrite to set the core shape. I think a little more docstring would be nice there, because otherwise it is hard to tell what that is for.

**Comment by ricardoV94** [comment]
> Add numba implementations for the random variables. (By using existing numba impls or adapting the numpy implementation and writing numba code)

Yup.

> Some minor changes to the numba rewrites: Adding rewrites for the inner graph of scans. (I thought we had a flag for ops that show if they have an inner graph and trigger them through that somehow? But I couldn't find that code)

We only trigger them at link time, has always been the case (except some rewrites that try to do stuff with existing Scans from the outside). Scan has an extra knob in that it allows users to specify it's inner compilation Mode (no other Ops with inner graphs allow this, I guess OFG indirectly, but it's not tested/documented). This is actually a source of trouble when we have a graph that may be compiled into distinct backends (happens in a typical PyMC workflow with external samplers).

We have to rethink the Scan having a user-customizable mode. Anyway this PR simply made sure we include the Numba required rewrite for RVs when dispatching the Scan impl.

> Change random variables so that they specify a signature. Maybe this should be mentioned in the release notes because it affects for instance pymc implementations or custom distributions?

This was not a breaking change, it will issue a warning telling people/PyMC to use signatures instead. More important imo to announce the actual Op node signature change and dtype is no longe part of it, and size=[] is no longer the same as size=None.
Will keep in mind to be verbose in the release notes

> Introduce a numba rewrite to set the core shape. I think a little more docstring would be nice there, because otherwise it is hard to tell what that is for.

Sure, will add

**Comment by ricardoV94** [comment]
Expanded the explanation of the core-shape numba rewrite

**Comment by ricardoV94** [comment]
@aseyboldt any suggestion about not suppressing all Numba Warnings :) The message match stopped working with the new `store_core_outputs` for some reason I can't tell (I tried with a bunch of strings/regexes without success)

**Comment by aseyboldt** [comment]
Why not just `message="Cannot cache compiled function"`? `filterwarnings` accepts a regex for the message, so maybe something in the string is interpreted in a strange way?

**Comment by ricardoV94** [comment]
I tried that, it did not work for some reason. If you want to try the relevant test is https://github.com/pymc-devs/pytensor/blob/1e96b8942bc42fd48af50e138c5b2458083c6091/tests/link/numba/test_basic.py#L1071-L1078

**Comment by aseyboldt** [comment]
Using this works for me:
```
warnings.filterwarnings(
    "ignore",
    message='Cannot cache compiled function "numba_funcified_fgraph"',
    category=NumbaWarning,
)

warnings.filterwarnings(
    "ignore",
    message='Cannot cache compiled function "store_core_outputs"',
    category=NumbaWarning,
)
```

**Comment by ricardoV94** [comment]
With that the test fails for me, maybe it's something add in my testing environment? I'll push and see what happens on the CI

**Comment by ricardoV94** [comment]
Also fails in the CI: https://github.com/pymc-devs/pytensor/actions/runs/9226914530/job/25387761898?pr=691#step:6:2878

**Comment by aseyboldt** [comment]
That's strange. If I run in for instance in the notebook, the warnings are in fact filtered.
Even if I put a clearly matching filterwarnings directly into the test function it doesn't work as long as the message isn't empty.
Could this be a problem with how we run the test? Does @pytest.mark.filterwarnings("error")  actually do what we want it to do?

**Comment by aseyboldt** [comment]
Probably that pytest function uses `warnings.filterwarnings` internally as well, and sets it to `error`. And maybe if there is a message the pytest filter runs first, and if there is no message in the filter our filter runs first?

**Comment by aseyboldt** [comment]
Maybe we could just use `with warnings.catch_warnings` in the test instead, and check that there were no warnings that way?

**Comment by ricardoV94** [comment]
> Maybe we could just use `with warnings.catch_warnings` in the test instead, and check that there were no warnings that way?

I'll give that a try

**Comment by ricardoV94** [comment]
Okay so pytest does override package defined filters: https://stackoverflow.com/questions/75345190/pytest-overrides-existing-warning-filters

However this leads me to think we should ignore this warning with pytest machinery, or else it will be emitted by every test that triggers the cache warning. 

If we do that I don't see how we can test the filter is working but maybe that's fine.

I am still puzzled as to why the filtering test was working before and no longer works now. But still does if we don't have a matching message

**Comment by ricardoV94** [comment]
Man what a stupid rabbit hole. So the numba message checking is failing because of NUMBA error message highlighting. The message fails to match in environments with `colorama` installed unless the environment flag `NUMBA_DISABLE_ERROR_MESSAGE_HIGHLIGHTING` is passed. 

https://github.com/numba/numba/blob/556545c5b2b162574c600490a855ba8856255154/numba/core/errors.py#L198-L202

This has nothing to do with pytest AFAICT. The messages are not filtered properly in an ipython cell either unless I set that flag as well. I couldn't figure out how to match a message with the bolding from colorama...

**Comment by ricardoV94** [comment]
Here is a test snippet:

```python
%env NUMBA_DISABLE_ERROR_MESSAGE_HIGHLIGHTING='1'
import colorama
import pytensor
import pytensor.tensor as pt
import numpy as np

x = pt.vector("x", shape=(5,), dtype="float64")
out = pt.psi(x) * 2
fn = pytensor.function([x], out, mode="NUMBA")

x_test = np.random.uniform(size=5)
fn(x_test)
```
Not setting that variable will lead it to fail (if colorama can be imported)

**Comment by ricardoV94** [comment]
Tests pass with https://github.com/pymc-devs/pytensor/pull/691/commits/a77df2406f995177b8398b597e933f1fb0767335

My only concern is that users with colorama will still see the warning. I don't know how to filter that kind of messages, and also agree that filtering any NumbaWarning is a tad too much. Should we perhaps ask Numba to have a subclass for caching warnings? Perhaps that would be more reasonable to filter by default?

**Comment by ricardoV94** [comment]
Ok fixed with some regex, I was being tripped by multiple bold escape codes...

**Comment by ricardoV94** [comment]
Marking as a draft because I want to merge a bugfix and do a release before this major changes get in

**Comment by aseyboldt** [comment]
Sounds like you had a fun afternoon.
But that explains the strange failures....

**Comment by twiecki** [comment]
Good one!

**Pull request merged**